### PR TITLE
feat(api): `nvim_open_tabpage`

### LINKED ISSUE #32893: lua equivalent of `win_new_tabpage`
### Problem

<details><summary>Final goal</summary>

To avoid an [XYProblem](https://xyproblem.info/), I'll start by saying my final goal:
I want something like `vim.lsp.buf.definition`, with some changes.

1. I want to it to open in a new tab and switch to it...
2. ...unless the file is already open in a tab, in which case I just want to switch to that tab.

I am trying to use `vim.lsp.ListOps.on_list`, but am running into the issue I'll describe now
</details>

In Lua, I am trying to open a new tab, create a window in that tab, and set the buffer for that window.  However, I am stuck on the first part.  I don't know how to open a new tab in Lua besides for just `tabnew`, but `tabnew` doesn't return the `tabnr`.

### Expected behavior

The C API has `‎win_new_tabpage‎`, which could work.  However, `‎win_new_tabpage‎` isn't available from the Lua API.

Would you accept a patch that adds `‎vim.api.nvim_win_new_tabpage‎`?

### PR DESCRIPTION
**Pull request by willothy**
## Problem
No programmatic interface (analogous to `nvim_open_win`) to open a new tabpage.

[use-cases](https://github.com/neovim/neovim/pull/27223#issuecomment-1963275909)

* `:tab sb bufnr` -> `nvim_open_tabpage(bufnr)`
* `:tabedit nvim_buf_get_name(bufnr)` -> `nvim_open_tabpage(bufnr)`
* `:tabedit new_name` -> `nvim_open_tabpage(fn.bufadd(new_name))`

## Solution
- Introduce `nvim_open_tabpage`. Benefits: 
    1. it returns the tabpage-id (unlke `:tabnew`)
    2. the `enter` parameter (similar to `nvim_open_win`)

fix https://github.com/neovim/neovim/issues/32893

### CODE REVIEW

File: src/nvim/api/tabpage.c [hunk 1]
```diff
@@ -14,6 +16,27 @@
 # include "api/tabpage.c.generated.h"
 #endif
 
+/// Opens a new tabpage with a single window
+///
+/// @param buffer   Buffer handle, or 0 for current buffer
+/// @param enter    Boolean, whether to enter the new tabpage
+/// @param[out] err Error details, if any
+/// @return Handle to newly created tabpage
+Tabpage nvim_open_tabpage(Buffer buffer, Boolean enter, Error *err)
```
Comment (justinmk) [inline src/nvim/api/tabpage.c:25]: Ignoring whether this API is wanted... we almost always want to accept `opt` and return `Dict`, unless we are 100% certain we won't need to expand the features in the future.
```suggestion
Dict nvim_open_tabpage(Buffer buffer, Boolean enter, Dict opts, Error *err)
```

File: src/nvim/window.c [hunk 2]
```diff
@@ -3972,28 +3972,31 @@ void free_tabpage(tabpage_T *tp)
 /// @param after Put new tabpage after tabpage "after", or after the current
 ///              tabpage in case of 0.
 /// @param filename Will be passed to apply_autocmds().
-/// @return Was the new tabpage created successfully? FAIL or OK.
-int win_new_tabpage(int after, char *filename)
+/// @return Pointer to the new tabpage, or NULL if error.
+tabpage_T *win_new_tabpage(int after, char *filename, bool enter)
```
Comment (justinmk) [inline src/nvim/window.c:3976]: The need for `enter` here is a sign that this API is more than just a wrapper, it's actually achieving new functionality. Right?

Comment (willothy) [reply src/nvim/window.c:3976]: Yes, since this would allow creating a new tabpage without entering it. Not sure if there's already a way to do that via Lua though, or if there's really even a use case for it.

File: src/nvim/api/tabpage.c (hunk 1 above)
Comment (willothy) [reply src/nvim/api/tabpage.c:25]: Good to know, I'll remember that in the future :)

re: whether this API is wanted

This was largely an experiment and I don't have a real use case for it, so no worries if it's not wanted.

**Comment by willothy** [comment]
I've added the "after" option and dict argument, and another test for it.

I'm not entirely sure why I need to add 1 to the tabpage index since `tabpage_index()` already returns a value, but it works...

File: src/nvim/api/tabpage.c [hunk 3]
```diff
@@ -14,6 +18,47 @@
...
+  tabpage_T *tp = win_new_tabpage(after, buf->b_ffname, enter);
+
+  if (!tp) {
+    api_set_error(err, kErrorTypeException, "Failed to create tabpage");
+    return 0;
+  }
+
+  win_set_buf(tp->tp_firstwin, buf, false, err);
+
+  return tp->handle;
```
Comment (seandewar) [inline src/nvim/api/tabpage.c:59]: Autocommands from `win_new_tabpage`, `win_set_buf` have to be considered here:

- `tp` may have been immediately freed (even `tp->handle` isn't valid in that case).
- `buf` may have been immediately freed.
- `tp->tp_firstwin` may be wrong if `win_new_tabpage` autocommands enter the tabpage and mess with the layout.

File: src/nvim/window.c [hunk 4]
```diff
@@ -4196,26 +4199,35 @@ int win_new_tabpage(int after, char *filename)
 
     newtp->tp_topframe = topframe;
     last_status(false);
-
     redraw_all_later(UPD_NOT_VALID);
```
Comment (seandewar) [inline src/nvim/window.c:4202]: The `!enter` handling in this function will need a more scrutinous review later.

For example, the window layout is corrupted with `!enter`, as this code has a number of assumptions that we're already in the new tabpage (e.g: `win_alloc_firstwin` overwrites the `topframe`, `firstwin` is changed, etc.).

File: src/nvim/window.c [hunk 5]
```diff
@@ -4196,26 +4199,35 @@ int win_new_tabpage(int after, char *filename)
...
     apply_autocmds(EVENT_WINNEW, NULL, NULL, false, curbuf);
-    apply_autocmds(EVENT_WINENTER, NULL, NULL, false, curbuf);
+    if (enter) {
+      apply_autocmds(EVENT_WINENTER, NULL, NULL, false, curbuf);
+    }
     apply_autocmds(EVENT_TABNEW, filename, filename, false, curbuf);
-    apply_autocmds(EVENT_TABENTER, NULL, NULL, false, curbuf);
+    if (enter) {
+      apply_autocmds(EVENT_TABENTER, NULL, NULL, false, curbuf);
+    }
```
Comment (seandewar) [inline src/nvim/window.c:4219]: With `!enter`, these autocommands won't fire within the context of the new tabpage, which maybe they should. Also, they could free `old_curwin`.

**Review by seandewar — COMMENTED** [review]
Here's a quick review from glancing at the code; I haven't yet tested this locally.
I probably won't be able to have a proper look until after 0.10.

**Comment by willothy** [comment]
These are great, thank you! Gonna mark this as draft again and keep working. 

After doing some testing (haven't worked on this branch in a little bit before today) I think I ran into a couple of those issues myself.

**Comment by justinmk** [comment]
@seandewar 's comments mentioned some important edge cases. I don't see `bdelete` or `bwipeout` in the tests. Do we have test coverage for "evil" autocmds that delete the window/tabpage during the event?

File: src/nvim/api/keysets_defs.h [hunk 6]
```diff
@@ -133,6 +133,11 @@ typedef struct {
   Boolean hide;
 } Dict(win_config);
 
+typedef struct {
+  OptionalKeys is_set__open_tabpage_opts_;
+  Integer after;
+} Dict(open_tabpage_opts);
```
Comment (justinmk) [inline src/nvim/api/keysets_defs.h:139]:
```suggestion
} Dict(open_tabpage);
```

File: src/nvim/api/keysets_defs.h [hunk 7]
```diff
@@ -133,6 +133,11 @@ typedef struct {
   Boolean hide;
 } Dict(win_config);
 
+typedef struct {
+  OptionalKeys is_set__open_tabpage_opts_;
```
Comment (justinmk) [inline src/nvim/api/keysets_defs.h:137]:
```suggestion
  OptionalKeys is_set__open_tabpage_;
```

File: src/nvim/api/tabpage.c [hunk 8]
```diff
@@ -15,6 +19,57 @@
...
+  if (!buf) {
+    return 0;
+  }
+  tabpage_T *tp = win_new_tabpage(after, buf->b_ffname, enter);
+  if (!tp) {
+    api_set_error(err, kErrorTypeException, "Failed to create tabpage");
+    return 0;
+  }
+
+  // TODO(willothy): What should happen if the window is immediately closed?
```
Comment (justinmk) [inline src/nvim/api/tabpage.c:56]: is the todo still open? looks like there is some handling  below

Comment (willothy) [reply src/nvim/api/tabpage.c:56]: Yep, forgot to remove the comment. Thanks!

**Comment by willothy** [comment]
> @seandewar 's comments mentioned some important edge cases. I don't see `bdelete` or `bwipeout` in the tests. Do we have test coverage for "evil" autocmds that delete the window/tabpage during the event?

There are checks in the code, but no coverage for that specifically - what should I check for? Just ensuring nvim doesn't crash or claim windows/tabs exist that don't?

**Review by zeertzjq — DISMISSED** [review]
`enter = false` is very broken:
- When the tabline was previously invisible, it does not become visible in both the old and the new tabpages, and text rendering is off by one row vs cursor position in the new tabpage.
- Null pointer dereference when trying to close the old tabpage.
- Heap-use-after-free when trying close to the new tabpage.
- Specifying a non-current buffer doesn't work.

**Comment by willothy** [comment]
Thank you, I'll get these fixed soon. Sorry about that, it's been a bit since I worked on this PR.

**Comment by vegerot** [comment]
unrelated question: is there an API to switch the current tab?  I don't see anything relevant on https://neovim.io/doc/user/api.html#api-tabpage.  If not, I'd be willing to contribute this feature after this PR lands.

([context](https://github.com/neovim/neovim/issues/32893#issuecomment-2725997744))

**Comment by seandewar** [comment]
@vegerot `:h nvim_set_current_tabpage()`

**Comment by vegerot** [comment]
Thanks!  The next API in the doc, `vim.api.nvim_get_current_win` is what I needed. 🚀

**Comment by willothy** [comment]
Finally got around to working on this again - I totally redid it with a separate function for when `enter=false` which I think simplified the implementation a lot. Also added the `after` field. 

Discussion points:
- imo the naming `after` is confusing since the new tabpage is actually placed *before* the tabpage of the provided number - `after` is really saying "put tabpage with id `after` after the new tabpage" which seems counterintuitive. Maybe we should just call it "index" instead?

**Comment by zeertzjq** [comment]
Hmm, the current implementation is almost the same as using `:tabnew` and then switching back to the original tabpage. It even still triggers the `TabLeave` and `TabEnter` autocommands, and the handling of `after` here is inconsistent with how `:tabnew` handles `[count]`. So is this really useful?

**Comment by willothy** [comment]
Ah yeah you're right definitely need to avoid triggering those autocmds since that is a large part of the point of this imo. As for `tabnew`, I definitely should have looked at that and will update this to match the numbering there. My numbering here was just based off of the internal window functions, which looking back on it doesn't make a lot of sense

If you still don't think it's worth it then we can close :)

File: src/nvim/window.c [hunk 9]
```diff
@@ -7565,3 +7555,89 @@ win_T *lastwin_nofloating(void)
...
+  if (win_alloc_firstwin(old_curtab->tp_curwin) == OK) {
+    // Insert the new tabpage in the correct position
+    if (after == 1) {
+      // New tab page becomes the first one
+      newtp->tp_next = first_tabpage;
+      first_tabpage = newtp;
+    } else {
+      tabpage_T *tp = old_curtab;
+      if (after > 0) {
+        // Put new tab page before tab page "after"
```
Comment (justinmk) [inline src/nvim/window.c]: I'm wondering why all this new code was needed, but existing code can't also make use of it? We need to stop duplicating stuff for different layers. There is already a way to open a tabpage, so there must be some way that code can be shared between this new way and the old way(s).

**Comment by justinmk** [comment]
## Status

- There is an [autocmds question](https://github.com/neovim/neovim/pull/27223#issuecomment-3315155946) which needs to be decided and will probably stall this until someone makes a call on it.
- the value of this new api is 
    1. it returns the tabpage-id (unlke `:tabnew`)
    2. the `enter` parameter
- [use-cases](https://github.com/neovim/neovim/pull/27223#issuecomment-1963275909)
  > * `:tab sb bufnr` -> `nvim_open_tabpage(bufnr)`
  > * `:tabedit nvim_buf_get_name(bufnr)` -> `nvim_open_tabpage(bufnr)`
  > * `:tabedit new_name` -> `nvim_open_tabpage(fn.bufadd(new_name))`

**Comment by seandewar** [comment]
This PR also has issues with validation, and can be refactored. I think I'll fix it up.

As for autocmds, I think this should only trigger `WinNew`, `TabNew`, plus those from `win_set_buf`, but maybe skipping Enter/Leave events except `BufWinEnter` (like `nvim_open_win`) if the window isn't entered?

**Comment by willothy** [comment]
Hey y'all, thanks for continuing the work on this! Happy to continue work on it to resolve the last few issues if that would be helpful :)

File: runtime/doc/api.txt [hunk 10]
```diff
@@ -3544,6 +3544,25 @@ nvim_set_option_value({name}, {value}, {opts})
...
+    Attributes: ~
+        not allowed when |textlock| is active
+        Since: 0.12.0
+
+    Parameters: ~
+      • {buffer}  (`integer`) Buffer to open in the first window of the new
+                  tabpage. Use 0 for current buffer.
+      • {config}  (`vim.api.keyset.tabpage_config`) Configuration for the new
+                  tabpage. Keys:
+                  • enter: Whether to enter the new tabpage (default: true)
```
Comment (justinmk) [inline runtime/doc/api.txt]: in `nvim_open_win()` this is a non-optional parameter. I guess it's more ergonomic to make it optional like this, but the inconsistency is a slight cost.

should the doc mention the implications for autocmds ?

Comment (seandewar) [reply runtime/doc/api.txt]: I don't mind making it non-optional if you'd like. Not sure which I prefer.

As for autocmds, unless someone complains later I think it's fine to exclude. It only triggers the `New` events with `!enter` (and other non-`Enter`/`Leave`, but can trigger `BufWinEnter`), which I think is intuitive.
It's like how `nvim_open_win` currently works with `!enter`, which doesn't mention anything special.

Comment (justinmk) [reply runtime/doc/api.txt]: I would lean towards consistency with `nvim_open_win`, unless there are strong objections.

**Review by justinmk — APPROVED** [review]
